src/hotspot/share/opto/loopopts.cpp

```diff
@@ -353,139 +353,161 @@ Node *PhaseIdealLoop::has_local_phi_input( Node *n ) {
+    IdealLoopTree *scale_loop = get_loop(scale_ctrl);
+    if (n_loop == scale_loop || !scale_loop->is_member(n_loop)) {
       return NULL;
-    const TypeInt *scale_t = scale->bottom_type()->isa_int();
-    if( scale_t && scale_t->is_con() && scale_t->get_con() >= 16 )
+    }
+    const TypeInt* scale_t = scale->bottom_type()->isa_int();
+    if (scale_t != NULL && scale_t->is_con() && scale_t->get_con() >= 16) {
       return NULL;              // Dont bother with byte/short masking
+    }
     // Add must vary with loop (else shift would be loop-invariant)
-    Node *add = n->in(1);
-    Node *add_ctrl = get_ctrl(add);
+    Node* add = n->in(1);
+    Node* add_ctrl = get_ctrl(add);
     IdealLoopTree *add_loop = get_loop(add_ctrl);
     //assert( n_loop == add_loop, "" );
-    if( n_loop != add_loop ) return NULL;  // happens w/ evil ZKM loops
+    if (n_loop != add_loop) return NULL;  // happens w/ evil ZKM loops
 
     // Convert I-V into I+ (0-V); same for V-I
-    if( add->Opcode() == Op_SubI &&
-        _igvn.type( add->in(1) ) != TypeInt::ZERO ) {
-      Node *zero = _igvn.intcon(0);
+    if (add->Opcode() == Op_Sub(bt) &&
+        _igvn.type(add->in(1)) != TypeInteger::zero(bt)) {
+      assert(add->Opcode() == Op_SubI || add->Opcode() == Op_SubL, "");
+      Node* zero = _igvn.integercon(0, bt);
       set_ctrl(zero, C->root());
-      Node *neg = new SubINode( _igvn.intcon(0), add->in(2) );
-      register_new_node( neg, get_ctrl(add->in(2) ) );
-      add = new AddINode( add->in(1), neg );
-      register_new_node( add, add_ctrl );
+      Node *neg = SubNode::make(zero, add->in(2), bt);
+      register_new_node(neg, get_ctrl(add->in(2)));
+      add = AddNode::make(add->in(1), neg, bt);
+      register_new_node(add, add_ctrl);
     }
-    if( add->Opcode() != Op_AddI ) return NULL;
+    if (add->Opcode() != Op_Add(bt)) return NULL;
+    assert(add->Opcode() == Op_AddI || add->Opcode() == Op_AddL, "");
     // See if one add input is loop invariant
     Node *add_var = add->in(1);
     Node *add_var_ctrl = get_ctrl(add_var);
-    IdealLoopTree *add_var_loop = get_loop(add_var_ctrl );
+    IdealLoopTree *add_var_loop = get_loop(add_var_ctrl);
     Node *add_invar = add->in(2);
     Node *add_invar_ctrl = get_ctrl(add_invar);
-    IdealLoopTree *add_invar_loop = get_loop(add_invar_ctrl );
-    if( add_var_loop == n_loop ) {
-    } else if( add_invar_loop == n_loop ) {
+    IdealLoopTree *add_invar_loop = get_loop(add_invar_ctrl);
+    if (add_var_loop == n_loop) {
+    } else if (add_invar_loop == n_loop) {
       // Swap to find the invariant part
       add_invar = add_var;
       add_invar_ctrl = add_var_ctrl;
       add_invar_loop = add_var_loop;
       add_var = add->in(2);
       Node *add_var_ctrl = get_ctrl(add_var);
-      IdealLoopTree *add_var_loop = get_loop(add_var_ctrl );
-    } else                      // Else neither input is loop invariant
+      IdealLoopTree *add_var_loop = get_loop(add_var_ctrl);
```

Comment:
The variable is unused. Is `get_loop()` just meant as sanity check? I first thought it accidentally shadows the previous variable definition on L396 instead of updating the variable but it is unused afterwards.

Reply:
Right, it has to be dead code.